### ISSUE TITLE
Schedule module

### ISSUE THREAD

**Issue by barbax7** [GrahamDumpleton/mod_wsgi#682]
Hi, i'm new user of apache and mod_wsgi
I've used waitress until now.

I've got some issues using schedule module: at first importing the flask app module, it executes the schedule functions, scheduling some tasks i need.

But sometimes, interacting with the webapp, i don't know why but it executes again schedule functions, so tasks are executed twice or more. I think that with some http requests it imports again the web app main file.

Is it normal? 

And, does the python script continue executing in background to execute scheduled tasks?

Thanks for your attenction.

**Comment by GrahamDumpleton** [comment]
Are you using embedded mode or daemon mode of mod_wsgi?

If using embedded mode it is more likely you will have multiple processes and as such your tasks can be run from different processes at different times. If the tasks are working against file system based or data store resources, this may not be desirable depending on your use case.

Next issue with embedded mode is that only the WSGI script file is reloaded if it is changed and not the whole process, so if your scheduled tasks are done as a side effect of importing the WSGI script file, they could be setup more than once in the life of the process. You would need to ensure that WSGI script file reloading it turned off if using embedded mode.

**Comment by barbax7** [comment]
Thanks for your reply.

Actually, in apache conf file i'm using DaemonMode

`WSGIDaemonProcess advicebook threads=5
WSGIScriptAlias / /var/www/advicebook/app.wsgi`

And in app.wsgi i import app as application from the main file.
Executing main file, it schedules the tasks i need.

Similar problem with another application, in this case i use, in the app.wsgi, the `from werkzeug.middleware.dispatcher import DispatcherMiddleware` to run more than one application.

**Comment by GrahamDumpleton** [comment]
That configuration alone isn't using daemon mode but would be using embedded mode.

Are you also setting ``WSGIProcessGroup``? That is required to have the WSGI application delegated to the daemon process group.

You can confirm what is happening for your configuration using the test in:

* https://modwsgi.readthedocs.io/en/master/user-guides/checking-your-installation.html#embedded-or-daemon-mode
* 
What do you get?

**Comment by barbax7** [comment]
Thanks again for your reply and attenction.

This is myconfiguration in Apache2

```
WSGIDaemonProcess start_all_flask threads=5
WSGIScriptAlias / /var/www/start_all_flask/app.wsgi

<Directory start_all_flask>
    WSGIProcessGroup start_all_flask
    WSGIApplicationGroup %{GLOBAL}
    Order deny,allow
    Allow from all
</Directory>
```

Chiecking my installation i receive `mod_wsgi.process_group = ''`


And, in wsgi file i import more than one app, from different packages. Well, sometimes the import occurs 1 second, and sometimes it occures about 20 seconds. Is it normal?

Thanks again for your help.

**Comment by GrahamDumpleton** [comment]
The problem with your configuration is that the argument to the Directory directive is wrong. It should be:

```
WSGIDaemonProcess start_all_flask threads=5
WSGIScriptAlias / /var/www/start_all_flask/app.wsgi

<Directory /var/www/start_all_flask>
    WSGIProcessGroup start_all_flask
    WSGIApplicationGroup %{GLOBAL}
    Order deny,allow
    Allow from all
</Directory>
```

Better still use:

```
WSGIDaemonProcess start_all_flask threads=5
WSGIScriptAlias / /var/www/start_all_flask/app.wsgi process-group=start_all_flask application-group=%{GLOBAL}

<Directory /var/www/start_all_flask>
    Order deny,allow
    Allow from all
</Directory>
```

as that avoids mucking up something like the Directory block.

The Order/Allow in that directory block appear to not really be doing anything, possibly because your Apache configuration is allowing access to the whole file system already, which is generally regarded as bad practice, or is allowing access to /var/www. There are some Apache configurations on distros which appear to have broken the Apache default setup of being secure by default which is a bit sad.

**Comment by barbax7** [comment]
In this way you suggested, will the python script be always alive like a Daemon?
Because i found out that lots of time i do a request, wsgi file takes from 20 to 30 seconds to import applications

I need to speed up

**Comment by GrahamDumpleton** [comment]
Yes. Read:

* https://modwsgi.readthedocs.io/en/master/user-guides/processes-and-threading.html

for the difference between embedded and daemon modes.

If it is taking 20-30 seconds to import your application then your application must be doing something during loading to cause that. Eg., calculating a data set or something.

If it is just that there is no capacity to handle requests because there are already too many requests being handled, it is because you don't have enough capacity.

As it is now, you can only handle 5 concurrent requests in total. If your requests are long running that could be a problem.

Before when you had the incorrect configuration Apache would have been scaling up the number of processes with a potential capacity of 150 concurrent requests if it was a typical default Apache configuration.

So how many requests/sec are you getting and how long do your requests take to run?

**Comment by barbax7** [comment]
I'll take a look at the docs.

Actually, while importing applications in wsgi file, they aren't running so many heavy functions. I mean, in one application there is only:
```
app = Flask(__name__)

@app.route('/')
def home_root():
    return render_template('index.html')
```
and it takes 6-7 seconds to import this one.
Then other 6-7 seconds to import a very little app, I use it for testing configurations.
And other 6-7 seconds for the principal app, that runs at start one API call to telegram and schedules 3 tasks.

It isn't normal.

There is also another Apache conf file for a site, where I have:

```
WSGIDaemonProcess advicebookthreads=5
WSGIScriptAlias / /var/www/advicebook/app.wsgi

<Directory advicebook>
    WSGIProcessGroup advicebook
    WSGIApplicationGroup %{GLOBAL}
    Order deny,allow
    Allow from all
</Directory>
```
Here, at the start, it schedules about 10 tasks and executes about 5 threads for update db and local variables.

**Comment by GrahamDumpleton** [comment]
That configuration for advice book is also likely wrong.

The argument to ``Directory`` should be an absolute file system path to the directory containing the WSGI script file.

Using a relative path of just "advicebook" is dangerous as don't know whether that will be interpreted as ``/advicebook``, or a sub directory with that name under the Apache server root, or just will not match at all ever as is not absolute path.

Change this as explained before for the other example to:

```
WSGIDaemonProcess advicebook threads=5
WSGIScriptAlias / /var/www/advicebook/app.wsgi process-group=advicebook application-group=%{GLOBAL}

<Directory /var/www/advicebook>
    Order deny,allow
    Allow from all
</Directory>
```

**Comment by barbax7** [comment]
I edited the config files. It looks like it is working better with scheduling, and also it seems that it doesn't execute the WSGI file at every HTTP call. So we've reached the goal.

If I edit the WSGI file, at the next HTTP request it is imported again. I think it's normal, but it takes about 20 seconds. I think I can execute a workflow that every time I edit the WSGI file I execute also an HTTP request to start the import. In this way, the webapp will be always available to users.

I'll let you know if there will be other issues. Thanks for your help!

**Comment by GrahamDumpleton** [comment]
Can you turn of automatic reloading when the WSGI script file changes so can better control when restart occurs.

* https://modwsgi.readthedocs.io/en/master/user-guides/reloading-source-code.html

**Comment by barbax7** [comment]
Hi, everything works, thanks to your help!

I have another question: when i was running the scripts on Heroku with waitress, responses were immediate.
Now, when i send a message to telegram bot, for example, it tooks 1 second to respond, even though the script is running in Daemon.

Any recommendations?

Somebody suggested me to use mod_python instead, but actually i don't like this solution.

**Comment by GrahamDumpleton** [comment]
The mod_python package is dead and no one should be using it due to the security risks with how it is implemented. I was the last one to work on mod_python in any meaningful way and mod_wsgi was written to replace it.

As to the delay, you need to confirm how long your WSGI request handler is taking to handle the request.

One option is to enable Apache access log and customise the log format to write out how long the request took to be handled.

A second option if using recent mod_wsgi is to add code to your application to log how long requests are taking. This latter one can give you better visibility in how long it takes and where time is being spent as it generates for each request a break down as to where in Apache request handling chain time is spent.

In simplest form, try adding to your WSGI script file the following code:

```
import mod_wsgi

def event_handler(name, **kwargs):
    if name == 'request_finished':
        print("REQUEST", kwargs)

mod_wsgi.subscribe_events(event_handler)
```

Provide an example of output where you believe it took 1 second.

**Comment by barbax7** [comment]
`[Sun May 30 09:05:25.799434 2021] [wsgi:error] [pid 72392] [remote 134.119.189.158:57358] REQUEST {'request_id': 'iyS1uyYO6SY', 'input_reads': 0, 'input_length': 0, 'input_time': 0.0, 'output_length': 178, 'output_writes': 1, 'cpu_user_time': 0.000361, 'cpu_system_time': 0.0, 'output_time': 1.9e-05, 'application_finish': 1622358325.799367, 'application_time': 0.00082, 'request_data': {}}`

`[Sun May 30 09:10:09.945998 2021] [wsgi:error] [pid 72392] [remote 91.108.6.88:42938] REQUEST {'request_id': 'fKyjzCYI6SY', 'input_reads': 2, 'input_length': 400, 'input_time': 9.1e-05, 'output_length': 1, 'output_writes': 1, 'cpu_user_time': 0.001787999999999998, 'cpu_system_time': 0.0, 'output_time': 0.077682, 'application_finish': 1622358609.945906, 'application_time': 0.07948, 'request_data': {}}`

These are the outputs after receiving and sending a message.

I can say also that at startup it turns up `System information disabled due to load higher than 1.0` and the commands `w` and `top` show that apache2 is using almost 100% of the CPU.
Could it be that the VPS is slow due to the high load?

**Comment by GrahamDumpleton** [comment]
For those two requests the time spent in the WSGI application was 0.8ms and 80ms. So way less than 1 second. The CPU user time is also very low, although accuracy for that isn't always great depending on the operating system.

Change now the code to:

```
import mod_wsgi

def event_handler(name, **kwargs):
    if name in ('request_started', 'request_finished'):
        print(name, kwargs)

mod_wsgi.subscribe_events(event_handler)
```

From details in the request started event will be able to determine if the blockage is within the Apache process when proxying.

**Comment by barbax7** [comment]
Sending message to bot:
`[Wed Jun 02 18:20:35.301219 2021] [wsgi:error] [pid 96402] [remote 91.108.6.88:37030] request_started {'request_id': '5/Ch1mr7P0g', 'thread_id': 2, 'daemon_connects': 1, 'daemon_restarts': 0, 'request_start': 1622650835.300583, 'queue_start': 1622650835.300745, 'daemon_start': 1622650835.300896, 'application_object': <werkzeug.middleware.dispatcher.DispatcherMiddleware object at 0x7f14485dbd30>, 'request_environ': {'GATEWAY_INTERFACE': 'CGI/1.1', 'SERVER_PROTOCOL': 'HTTP/1.1', 'REQUEST_METHOD': 'POST', 'QUERY_STRING': '', 'REQUEST_URI': '/opentgarchive/1765187416:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'SCRIPT_NAME': '', 'PATH_INFO': '/opentgarchive/1765187416:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'PATH_TRANSLATED': '/var/www/start_all_flask/app.wsgi/opentgarchive/1765187416:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'HTTP_HOST': '18.224.96.81', 'CONTENT_TYPE': 'application/json', 'CONTENT_LENGTH': '400', 'HTTP_CONNECTION': 'keep-alive', 'HTTP_ACCEPT_ENCODING': 'gzip, deflate', 'SERVER_SIGNATURE': '<address>Apache/2.4.41 (Ubuntu) Server at 18.224.96.81 Port 443</address>\\n', 'SERVER_SOFTWARE': 'Apache/2.4.41 (Ubuntu)', 'SERVER_NAME': '18.224.96.81', 'SERVER_ADDR': '172.31.45.222', 'SERVER_PORT': '443', 'REMOTE_ADDR': '91.108.6.88', 'DOCUMENT_ROOT': '/var/www/', 'REQUEST_SCHEME': 'https', 'CONTEXT_PREFIX': '', 'CONTEXT_DOCUMENT_ROOT': '/var/www/', 'SERVER_ADMIN': 'webmaster@localhost', 'SCRIPT_FILENAME': '/var/www/start_all_flask/app.wsgi', 'REMOTE_PORT': '37030', 'mod_wsgi.script_name': '', 'mod_wsgi.path_info': '/opentgarchive/1765187416:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'mod_wsgi.process_group': 'start_all_flask', 'mod_wsgi.application_group': '', 'mod_wsgi.callable_object': 'application', 'mod_wsgi.request_handler': 'wsgi-script', 'mod_wsgi.handler_script': '', 'mod_wsgi.script_reloading': '1', 'mod_wsgi.listener_host': '', 'mod_wsgi.listener_port': '443', 'mod_wsgi.enable_sendfile': '0', 'mod_wsgi.ignore_activity': '0', 'mod_wsgi.request_start': '1622650835300583', 'mod_wsgi.request_id': '5/Ch1mr7P0g', 'mod_wsgi.queue_start': '1622650835300745', 'mod_wsgi.daemon_connects': '1', 'mod_wsgi.daemon_restarts': '0', 'mod_wsgi.daemon_start': '1622650835300896', 'mod_wsgi.script_start': '1622650835301003', 'wsgi.version': (1, 0), 'wsgi.multithread': True, 'wsgi.multiprocess': False, 'wsgi.run_once': False, 'wsgi.url_scheme': 'https', 'wsgi.errors': <_io.TextIOWrapper name='<wsgi.errors>' encoding='utf-8'>, 'wsgi.input': <mod_wsgi.Input object at 0x7f1444c3a7a0>, 'wsgi.input_terminated': True, 'wsgi.file_wrapper': <class 'mod_wsgi.FileWrapper'>, 'apache.version': (2, 4, 41), 'mod_wsgi.version': (4, 6, 8), 'mod_wsgi.total_requests': 1, 'mod_wsgi.thread_id': 2, 'mod_wsgi.thread_requests': 0}, 'application_start': 1622650835.301003, 'request_data': {}}`

`[Wed Jun 02 18:20:35.380798 2021] [wsgi:error] [pid 96402] [remote 91.108.6.88:37030] request_finished {'request_id': '5/Ch1mr7P0g', 'input_reads': 2, 'input_length': 400, 'input_time': 8.8e-05, 'output_length': 1, 'output_writes': 1, 'cpu_user_time': 0.001562, 'cpu_system_time': 0.0, 'output_time': 0.078137, 'application_finish': 1622650835.380709, 'application_time': 0.079706, 'request_data': {}}`

**Comment by barbax7** [comment]
Sending message to bot:
`[Wed Jun 02 18:20:35.301219 2021] [wsgi:error] [pid 96402] [remote 91.108.6.88:37030] request_started {'request_id': '5/Ch1mr7P0g', 'thread_id': 2, 'daemon_connects': 1, 'daemon_restarts': 0, 'request_start': 1622650835.300583, 'queue_start': 1622650835.300745, 'daemon_start': 1622650835.300896, 'application_object': <werkzeug.middleware.dispatcher.DispatcherMiddleware object at 0x7f14485dbd30>, 'request_environ': {'GATEWAY_INTERFACE': 'CGI/1.1', 'SERVER_PROTOCOL': 'HTTP/1.1', 'REQUEST_METHOD': 'POST', 'QUERY_STRING': '', 'REQUEST_URI': '/opentgarchive/***:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'SCRIPT_NAME': '', 'PATH_INFO': '/opentgarchive/***:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'PATH_TRANSLATED': '/var/www/start_all_flask/app.wsgi/opentgarchive/***:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'HTTP_HOST': '18.224.96.81', 'CONTENT_TYPE': 'application/json', 'CONTENT_LENGTH': '400', 'HTTP_CONNECTION': 'keep-alive', 'HTTP_ACCEPT_ENCODING': 'gzip, deflate', 'SERVER_SIGNATURE': '<address>Apache/2.4.41 (Ubuntu) Server at 18.224.96.81 Port 443</address>\\n', 'SERVER_SOFTWARE': 'Apache/2.4.41 (Ubuntu)', 'SERVER_NAME': '18.224.96.81', 'SERVER_ADDR': '172.31.45.222', 'SERVER_PORT': '443', 'REMOTE_ADDR': '91.108.6.88', 'DOCUMENT_ROOT': '/var/www/', 'REQUEST_SCHEME': 'https', 'CONTEXT_PREFIX': '', 'CONTEXT_DOCUMENT_ROOT': '/var/www/', 'SERVER_ADMIN': 'webmaster@localhost', 'SCRIPT_FILENAME': '/var/www/start_all_flask/app.wsgi', 'REMOTE_PORT': '37030', 'mod_wsgi.script_name': '', 'mod_wsgi.path_info': '/opentgarchive/***:AAGPlojNrWHffbuT4wSs6pqMI2ZIyRecKgM', 'mod_wsgi.process_group': 'start_all_flask', 'mod_wsgi.application_group': '', 'mod_wsgi.callable_object': 'application', 'mod_wsgi.request_handler': 'wsgi-script', 'mod_wsgi.handler_script': '', 'mod_wsgi.script_reloading': '1', 'mod_wsgi.listener_host': '', 'mod_wsgi.listener_port': '443', 'mod_wsgi.enable_sendfile': '0', 'mod_wsgi.ignore_activity': '0', 'mod_wsgi.request_start': '1622650835300583', 'mod_wsgi.request_id': '5/Ch1mr7P0g', 'mod_wsgi.queue_start': '1622650835300745', 'mod_wsgi.daemon_connects': '1', 'mod_wsgi.daemon_restarts': '0', 'mod_wsgi.daemon_start': '1622650835300896', 'mod_wsgi.script_start': '1622650835301003', 'wsgi.version': (1, 0), 'wsgi.multithread': True, 'wsgi.multiprocess': False, 'wsgi.run_once': False, 'wsgi.url_scheme': 'https', 'wsgi.errors': <_io.TextIOWrapper name='<wsgi.errors>' encoding='utf-8'>, 'wsgi.input': <mod_wsgi.Input object at 0x7f1444c3a7a0>, 'wsgi.input_terminated': True, 'wsgi.file_wrapper': <class 'mod_wsgi.FileWrapper'>, 'apache.version': (2, 4, 41), 'mod_wsgi.version': (4, 6, 8), 'mod_wsgi.total_requests': 1, 'mod_wsgi.thread_id': 2, 'mod_wsgi.thread_requests': 0}, 'application_start': 1622650835.301003, 'request_data': {}}`

`[Wed Jun 02 18:20:35.380798 2021] [wsgi:error] [pid 96402] [remote 91.108.6.88:37030] request_finished {'request_id': '5/Ch1mr7P0g', 'input_reads': 2, 'input_length': 400, 'input_time': 8.8e-05, 'output_length': 1, 'output_writes': 1, 'cpu_user_time': 0.001562, 'cpu_system_time': 0.0, 'output_time': 0.078137, 'application_finish': 1622650835.380709, 'application_time': 0.079706, 'request_data': {}}`

**Comment by GrahamDumpleton** [comment]
Key values are:

```
'request_start': 1622650835.300583
'queue_start': 1622650835.300745
'daemon_start': 1622650835.300896
'application_start': 1622650835.301003
'application_time': 0.079706
```

The request start is when Apache first accepted the request.

The queue start is when Apache child worker process determined that the request needs to be proxied to the mod_wsgi daemon process and starts attempting to connect to the daemon process group socket.

The daemon start is when the mod_wsgi daemon process accepted the request.

The application start is when the mod_wsgi daemon process handed it off to the WSGI application.

The difference between application start and request start was:

```
0.0004200935363769531
```

This means that it only took 0.4ms between Apache accepting the request and it being handed off to the WSGI application. Subsequently the application time, which is how long the WSGI application took to handle the request and write back the response from the mod_wsgi daemon process, was:

```
0.079706
```

or 80ms.

In other words there was no delay inside of Apache which can account for long delays.

Note that these timings don't take into consideration how long a request sat in the Apache listener socket queue, but there shouldn't be any delay there unless your whole Apache server was getting a huge number of concurrent connections and it was overloaded.

So all this indicates that delays are occurring somewhere outside of Apache.

Because of the various issues you have had along the way, can you clarify again exactly what the problem is you are trying to track down at the moment so I am not looking in the wrong direction as to a reason.

**Comment by GrahamDumpleton** [comment]
Closing as no further response and assuming enough information provided to keep going.
